examples/experimental/script_to_assets_branching_io_manager/hn_dagster_external_execution.py

```diff
@@ -0,0 +1,140 @@
+import json
+import subprocess
+import sys
+from typing import List, Optional, Sequence, Set
+
+from dagster import (
+    AssetsDefinition,
+    Definitions,
+    Nothing,
+    OpExecutionContext,
+    asset,
+    define_asset_job,
+)
+from dagster._core.definitions.events import AssetKey
+from dagster._core.instance import DagsterInstance
+from dagster._core.storage.branching.branching_io_manager import (
+    get_text_metadata_value,
+    latest_materialization_log_entry,
+)
+from dagster._utils import file_relative_path
+
+STORAGE_NAMESPACE_METADATA_KEY = "storage_namespace"
+
+
+def upstream_asset_partition_materialized_in_storage_namespace(
+    instance: DagsterInstance,
+    upstream_asset_key: AssetKey,
+    partition_key: Optional[str],
+    storage_namespace: str,
+):
+    event_log_entry = latest_materialization_log_entry(
+        instance=instance,
+        asset_key=upstream_asset_key,
+        partition_key=partition_key,
+    )
+    return (
+        event_log_entry
+        and event_log_entry.asset_materialization
+        and get_text_metadata_value(
+            event_log_entry.asset_materialization, STORAGE_NAMESPACE_METADATA_KEY
+        )
+        == storage_namespace
+    )
+
+
+def in_branch():
+    return True
+
+
+def create_asset_namespace_mapping(
+    instance: DagsterInstance,
+    current_asset_key_str: str,
+    partition_key: Optional[str],
+    upstream_asset_array: List[str],
+    parent_storage_namespace: str,
+    branch_storage_namespace: str,
+    in_branch: bool,
+):
+    if not in_branch:
+        return {
+            asset_key_str: parent_storage_namespace
+            for asset_key_str in [current_asset_key_str, *upstream_asset_array]
+        }
+
+    # always write this asset to branch namespace if in branch
+    asset_namespaces = {current_asset_key_str: branch_storage_namespace}
+
+    # then for every upstream asset see if it has been materialized in the branch
+    # if not, read from the parent
+    for upstream_asset in upstream_asset_array:
+        upstream_asset_key = AssetKey(upstream_asset)
+        if upstream_asset_partition_materialized_in_storage_namespace(
+            instance=instance,
+            upstream_asset_key=upstream_asset_key,
+            partition_key=partition_key,
+            storage_namespace=branch_storage_namespace,
+        ):
+            asset_namespaces[upstream_asset] = branch_storage_namespace
+        else:
+            asset_namespaces[upstream_asset] = parent_storage_namespace
+
+    return asset_namespaces
```

Comment:
This function queries the event log in order to determine whether or not to load from the parent (usually prod) or the branch environment